src/awe/rmdparchive.cpp

```diff
@@ -196,91 +185,83 @@ Common::ReadStream *RMDPArchive::getResource(const std::string &rid) const {
 }
 
 bool RMDPArchive::hasResource(const std::string &rid) const {
-	std::stringstream path(
-			std::regex_replace(
-					(_pathPrefix ? "d:/data/" : "") + rid,
-					std::regex("\\\\"),
-					"/"
-			)
-	);
-	std::string item;
-
-	FolderEntry folder = _folderEntries.front();
-
-	uint32_t nameHash;
-
-	while (std::getline(path, item, '/')) {
-		nameHash = Common::crc32(Common::toLower(item));
+	std::stringstream path = this->getNormalizedPath(rid);
 
-		if (path.eof())
-			break;
+	auto maybeFolder = this->findDirectory(path);
+	if (!maybeFolder.has_value()) return false;
+	FolderEntry folder = maybeFolder.value();
 
-		if (folder.nextLowerFolder == 0xFFFFFFFF)
-			return false;
+	auto maybeFile = this->findFile(folder, folder.nameHash);
+	return maybeFile.has_value();
+}
 
-		folder = _folderEntries[folder.nextLowerFolder];
+bool RMDPArchive::hasDirectory(const std::string &directory) const {
+	std::stringstream path = this->getNormalizedPath(directory);
+	auto maybeFolder = this->findDirectory(path);
+	return maybeFolder.has_value();
+}
 
-		while (nameHash != folder.nameHash) {
-			if (folder.nextNeighbourFolder == 0xFFFFFFFF)
-				return false;
-			folder = _folderEntries[folder.nextNeighbourFolder];
-		}
+std::string RMDPArchive::readEntryName(Common::ReadStream *bin, uint32_t offset, uint32_t nameSize) {
+	std::string result;
+	if (offset != NO_ENTRY) {
+			size_t lastPos = bin->pos();
+			bin->seek(-static_cast<int>(nameSize) + static_cast<int>(offset), Common::ReadStream::END);
+			result = bin->readNullTerminatedString();
+			bin->seek(lastPos);
+	} else {
+		result = "";
 	}
+	return result;
+}
 
-	if (folder.nextFile == 0xFFFFFFFF)
-		return false;
+RMDPArchive::FolderEntry RMDPArchive::readFolder(Common::ReadStream *bin, uint32_t (Common::ReadStream::*readUint32) (), uint32_t nameSize) {
+	// Since folder struct remains the same through v2/7/8 headers
+	FolderEntry entry;
+	entry.nameHash = (bin->*readUint32)();
+	entry.nextNeighbourFolder = (bin->*readUint32)();
+	entry.prevFolder = (bin->*readUint32)();
 
-	FileEntry file = _fileEntries[folder.nextFile];
+	bin->skip(4); // Unknown value
 
-	while (file.nameHash != nameHash) {
-		if (file.nextFile == 0xFFFFFFFF)
-			return false;
-		file = _fileEntries[file.nextFile];
-	}
+	uint32_t nameOffset = (bin->*readUint32)();
+	entry.name = this->readEntryName(bin, nameOffset, nameSize);
 
-	return true;
+	entry.nextLowerFolder = (bin->*readUint32)();
+	entry.nextFile = (bin->*readUint32)();
+	return entry;
 }
 
-bool RMDPArchive::hasDirectory(const std::string &directory) const {
-	std::stringstream path(
-		std::regex_replace(
-			(_pathPrefix ? "d:/data/" : "") + directory,
-			std::regex("\\\\"),
-			"/"
-		)
-	);
-	std::string item;
-
-	FolderEntry folder = _folderEntries.front();
-
-	uint32_t nameHash;
+RMDPArchive::FileEntry RMDPArchive::readFile(Common::ReadStream *bin, uint32_t (Common::ReadStream::*readUint32) (), uint64_t (Common::ReadStream::*readUint64) (), uint32_t nameSize) {
```

Comment:
As much as I like the idea of this method, I fear you have to add a version parameter since the format will change slightly between versions of the rmdp format. For example american nightmare added a 64 bit timestamp or quantum break extended offset and size to 64 bit.

Reply:
As an option, it can be renamed to `readFileBasicInfo` and changed to read only version-agnostic information, since a lot of fields are similar anyway.

Reply:
I would prefer the version parameter as I explained in a later to keep this the reading of the file values together.